File: src/main/java/com/gandan/practice/trycatch/TryCatchFinally.java
package com.gandan.practice.trycatch;

import com.gandan.practice.task.ITask;

/**
 * Return in try, catch, and finally.
 *
 * Created by chandra on 12/3/15.
 */
public class TryCatchFinally implements ITask {
    @Override
    public void execute() {
        print();
    }

    private void print() {
        System.out.println("Return in Try-Finally result in " + returnInTryAndFinally());
        System.out.println("Return in Catch-Finally result in " + returnInTryAndFinally());
    }

    private String returnInTryAndFinally() {
        try {
            return "Try clause";
        } finally {
            return "Finally clause";
        }
    }

    private String returnInCatchAndFinally() {
        try {
            throw new RuntimeException("simulate exception");
        } catch (Exception e) {
            return "Catch Clause";
        } finally {
            return "Finally Clause";
        }

    }
}

File: src/main/kotlin/com/gandan/practice/trycatch/TryCatchFinally.kt
package com.gandan.practice.trycatch

fun main(args: Array<String>) {
    TryCatchFinally().execute()
}

class TryCatchFinally {
    fun execute() {
        print()
    }

    private fun print() {
        println("Return in Try-Finally result in " + returnInTryAndFinally())
        println("Return in Catch-Finally result in " + returnInCatchAndFinally())
    }

    @Suppress("UNREACHABLE_CODE")
    private fun returnInTryAndFinally(): String {
        try {
            return "Try clause"
        } finally {
            return "Finally clause"
        }
    }

    @Suppress("UNREACHABLE_CODE")
    private fun returnInCatchAndFinally(): String {
        try {
            throw RuntimeException("simulate exception")
        } catch (e: Exception) {
            return "Catch Clause"
        } finally {
            return "Finally Clause"
        }

    }
}
